gradio/blocks.py

```diff
@@ -614,19 +614,47 @@ def preprocess_data(self, fn_index, raw_input, state):
             processed_input = raw_input
         return processed_input
 
-    async def call_function(self, fn_index, processed_input):
+    async def call_function(self, fn_index, processed_input, iterator=None):
         """Calls and times function with given index and preprocessed input."""
         block_fn = self.fns[fn_index]
-
+        is_generating = False
         start = time.time()
-        if inspect.iscoroutinefunction(block_fn.fn):
-            prediction = await block_fn.fn(*processed_input)
-        else:
-            prediction = await anyio.to_thread.run_sync(
-                block_fn.fn, *processed_input, limiter=self.limiter
-            )
+
+        if iterator is None:  # If not a generator function that has already run
+            if inspect.iscoroutinefunction(block_fn.fn):
+                prediction = await block_fn.fn(*processed_input)
+            else:
+                prediction = await anyio.to_thread.run_sync(
+                    block_fn.fn, *processed_input, limiter=self.limiter
+                )
+
+        if inspect.isasyncgenfunction(block_fn.fn):
+            raise ValueError("Gradio does not support async generators.")
+        if inspect.isgeneratorfunction(block_fn.fn):
+            if not self.enable_queue:
+                raise ValueError("Need to enable queue to use generators.")
+            try:
+                if iterator is None:
+                    iterator = prediction
+                prediction = next(iterator)
```

Comment:
Another weird quirk compared to stateful-non iterators is that you technically don't have to return state from your function and it will still work:

```python
import gradio as gr
import time

def add(a, b):
    return a + b


def cumulative_sum(up_to, current=0):
    for i in range(up_to):
        time.sleep(0.5)
        current = add(i, current)
        yield current


with gr.Blocks() as demo:

    with gr.Row():
        with gr.Column():
            num = gr.Number(precision=0)
            state = gr.State(value=0)
            btn = gr.Button("Count")
        with gr.Column():
            output = gr.Number()
        
        btn.click(lambda: 0, inputs=None, outputs=state)
        btn.click(cumulative_sum, inputs=[num, state], outputs=[output])

demo.queue().launch()
```

Reply:
So this demo actually works for me:

```py
import gradio as gr
import time

def cumulative_sum(up_to, current_sum=0):
    for i in range(up_to):
        time.sleep(0.5)
        new = current_sum + i
        yield new, new

gr.Interface(cumulative_sum, inputs=[gr.Number(precision=0), "state"],
             outputs=[gr.Number(), "state"]).queue().launch()
```

Does it not work for you?

Reply:
@abidlabs It works but it's not actually doing the cumulative sum. 

![rename_state_in_iterator](https://user-images.githubusercontent.com/41651716/189138494-a37a9d0d-2e90-4fad-a779-928da480c3b5.gif)

The reason is that at each iteration of the loop, the value of `current_sum` is 0 since the function is not called again with the new state value. The fact the function is only called once is why you technically don't need to return state with an iterator. Users may expect their function to be called multiple times but that's not actually what's happening.

I'm not sure if there is a way around this with the `yield` iterator api. After thinking about it yesterday, I think it's ok. We can introduce a non-breaking additional api for iterative outputs if users find this confusing. I do think we should eventually mention this in the docs.